File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.java

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Scanner;

public class Main {

	static class Edge implements Comparable<Edge> {
		int A;
		int B;
		int W;

		public Edge(int a, int b, int w) {
			super();
			A = a;
			B = b;
			W = w;
		}

		@Override
		public String toString() {
			return "Edge [A=" + A + ", B=" + B + ", W=" + W + "]";
		}

		@Override
		public int compareTo(Edge o) {
			return Integer.compare(this.W, o.W);
		}

	}

	public static void main(String[] args) {

		// 최소신장트리
		// 1. 섬 갯수도 구해야함.
		// 2. 섬마다의 가중치를 모두 구한다.
		// 3. 가장 최소로 되는 모든 다리의 합.
		// * 다리의 길이는 2이상!! (사각형들의 끝 값에서 2이상이 되야함)

		Scanner sc = new Scanner(System.in);

		edgelist = new ArrayList<>();


		n = sc.nextInt();
		m = sc.nextInt();

		grid = new int[n][m];
		visited = new boolean[n][m];

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				grid[i][j] = sc.nextInt();
			}
		}
		islandnum = 0;
		int numbering = 1;
		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (grid[i][j] == 1 && !visited[i][j]) {
					dfs(i, j, numbering);
					numbering++; // 섬 넘버링
					islandnum++; // 섬의 갯수
				}
			}
		}

//		for (int i = 0; i < n; i++) {
//			for (int j = 0; j < m; j++) {
//				System.out.print(grid[i][j] + " ");
//			}
//			System.out.println();
//		}

		visited = new boolean[n][m];

//		System.out.println(Arrays.deepToString(checkunique));

		for (int i = 0; i < n; i++) {
			for (int j = 0; j < m; j++) {
				if (grid[i][j] != 0) {
					caldist(i, j, grid[i][j]);
				}
			}
		}
//		System.out.println(edgelist);

		Collections.sort(edgelist);
		parents = new int[islandnum + 1];

		for (int i = 0; i < islandnum + 1; i++) {
			parents[i] = i;
		}

		int cnt = 0;
		int sum = 0;
//		System.out.println(edgelist);
		for (int i = 0; i < edgelist.size(); i++) {
			int ap = findSet(edgelist.get(i).A);
			int bp = findSet(edgelist.get(i).B);

			if (ap != bp) {
				union(ap, bp);
				sum += edgelist.get(i).W;
				cnt++;
			}
			if (cnt == islandnum - 1) {
				break;
			}

		}

//		System.out.println(Arrays.toString(parents));

		int root = findSet(1);
		for (int i = 2; i < parents.length; i++) {
			if (findSet(i) != root) {
				sum = -1;
				break;
			}
		}
		System.out.println(sum);

	}

	private static void union(int ap, int bp) {

		if (ap != bp) {
			parents[bp] = ap;
		}

	}

	private static int findSet(int x) {
		if (parents[x] != x) {
			parents[x] = findSet(parents[x]);
		}
		return parents[x];
	}

	static int[] parents;
	static int[] costs;

	static int[] row = { 1, -1, 0, 0 };
	static int[] col = { 0, 0, 1, -1 };
	static int n;
	static int m;
	static int[][] grid;
	static boolean[][] visited;
	static int idx = 0;
	static List<Edge> edgelist;
	static int islandnum;

	private static void caldist(int r, int c, int from) {
		for (int k = 0; k < 4; k++) {
			int nr = r + row[k];
			int nc = c + col[k];
			int dist = 0;
			while (nr >= 0 && nr < n && nc >= 0 && nc < m && grid[nr][nc] != from) {
				if (nr >= 0 && nr < n && nc >= 0 && nc < m && grid[nr][nc] != 0) {
					int to = grid[nr][nc];
					if (dist >= 2 && from < to) {
						edgelist.add(new Edge(from, to, dist));
					}
					break;
				}
				nr += row[k];
				nc += col[k];
				dist++;
			}
		}
	}

	private static void dfs(int r, int c, int numbering) {
		visited[r][c] = true;
		grid[r][c] = numbering;

		for (int k = 0; k < 4; k++) {
			int nr = r + row[k];
			int nc = c + col[k];
			if (nr >= 0 && nr < n && nc >= 0 && nc < m && !visited[nr][nc] && grid[nr][nc] == 1) {
				visited[nr][nc] = true;
				grid[r][c] = numbering;
				dfs(nr, nc, numbering);
			}
		}

	}

}

File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.py
'''
섬들을 넘버링하고
섬들의 테두리에서 가로/ 세로로 뻗었을 때 다른 섬에 닿으면 그 길이를 기록. 그담에 연결할 수 있는 최소..를 ㄻ너우마움니움
다익스트라?
'''
from collections import deque

n, m = map(int, input().split())
grid = [list(map(int, input().split())) for i in range(n)]

visited = [[0] * m for i in range(n)]

num = 1
row = [-1, 1, 0, 0]
col = [0, 0, 1, -1]


def bfs(r, c):
    q = deque([(r, c)])
    while q:
        r, c = q.popleft()
        for k in range(4):
            nr = r + row[k]
            nc = c + col[k]
            if not (0 <= nr < n and 0 <= nc < m) or not grid[nr][nc] or visited[nr][nc]:
                continue
            visited[nr][nc] = num
            q.append((nr, nc))


for i in range(n):
    for j in range(m):
        if grid[i][j] and not visited[i][j]:
            visited[i][j] = num
            bfs(i, j)
            num += 1


adj = set()  # 정점 번호와 길이 .

for i in range(n):
    for j in range(m):
        if visited[i][j]:
            mosuri = False
            for k in range(4):
                nr = i + row[k]
                nc = j + col[k]
                if not (0 <= nr < n and 0 <= nc < m):
                    continue
                if grid[nr][nc] == 0:
                    mosuri = True
                    break
            if mosuri:
                for k in range(4):
                    r, c = i, j
                    dist = 0
                    while True:
                        nr = r + row[k]
                        nc = c + col[k]
                        if not (0 <= nr < n and 0 <= nc < m) or visited[nr][nc] == visited[i][j]:
                            break
                        dist += 1
                        if visited[nr][nc]:
                            if dist > 2:
                                tmp = [visited[i][j], visited[nr][nc]]
                                tmp.sort()
                                adj.add((tmp[0], tmp[1], dist - 1))
                            break
                        r = nr
                        c = nc

adj = list(adj)
adj.sort(key=lambda x: x[2])


p = [-1] * num
for i in range(1, num):
    p[i] = i


cnt, sum = 0, 0


def find(x):
    if p[x] != x:
        p[x] = find(p[x])
    return p[x]


def union(ap, bp):
    if ap != bp:
        p[bp] = ap


for a,b,dist in adj:
    ap = find(a)
    bp = find(b)
    if ap != bp:
        union(ap, bp)
        sum += dist
        cnt += 1
    if cnt == num - 1:
        break

myp = find(1)

for you in range(2,num):
    youp = find(you)
    if myp != youp:
        sum = -1
        break
print(sum)
